### PR TITLE
Oracle Support - Add CPI and CSI reconciler for Oracle Cloud to TKG Addons Controller

### PR DESCRIPTION
**Pull request by HanFa**
Co-authored-by: Fang Han <fhan@vmware.com>
Co-authored-by: Xudong Liu <xudongl@vmware.com>
Signed-off-by: Naadir Jeewa <jeewan@vmware.com>

### What this PR does / why we need it

Required by and cherry-picked from the Oracle Support, https://github.com/vmware-tanzu/tanzu-framework/pull/3557, documented in https://github.com/vmware-tanzu/tanzu-framework/pull/3600

Add Oracle CPI package, CR and its CPIConfig controller.

### Which issue(s) this PR fixes

Fixes #

### Release note

```release-note
Oracle Support - Add CPI and CSI reconciler for Oracle Cloud to TKG Addons Controller
```

### CODE REVIEW

**Comment by randomvariable** [comment]
Some build errors on this PR.

**Comment by HanFa** [comment]
Fixed the addon test. retriggering CI

File: addons/controllers/cpi/oraclecpiconfig_controller.go [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+			Subnet1 string `yaml:"subnet1"`
+			Subnet2 string `yaml:"subnet2"`
+		}{
+			Subnet1: subnet,
+			Subnet2: subnet,
+		},
+	}
+
+	// generate data value secrets for the Oracle CPI package
+	secret := &v1.Secret{
```
Comment (XudongLiuHarold) [inline addons/controllers/cpi/oraclecpiconfig_controller.go:187]: nit: You can assemble secret object here, which will make the code more readable

Comment (XudongLiuHarold) [reply addons/controllers/cpi/oraclecpiconfig_controller.go:187]: stringData, ownerReference etc.

File: addons/controllers/cpi/oraclecpiconfig_util.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+	}
+
+	r.Log.V(4).Info("Mapping Cluster to OracleCPIConfig")
+
+	cs := &cpiv1alpha1.OracleCPIConfigList{}
+	_ = r.List(context.Background(), cs)
+	var requests []ctrl.Request
+
+	for _, cpiConfig := range cs.Items {
+		requests = forClusterMappedByCPIConfigEnqueueRequest(&cpiConfig, cluster, r.Config.ConfigControllerConfig, r.Log, requests)
```
Comment (XudongLiuHarold) [inline addons/controllers/cpi/oraclecpiconfig_util.go:35]: Shouldn't we only trigger reconcile for this cluster's oracle cpi config?

We can check each OracleCPIConfig object, to see if they have owner reference to this cluster, if not, we should skip it.

File: addons/controllers/cpi/vspherecpiconfig_utils.go [hunk 3]
```diff
@@ -23,14 +25,64 @@ import (
...
+
+	// corresponding CPIConfig should have following ownerRef
+	ownerReference := metav1.OwnerReference{
+		APIVersion: clusterapiv1beta1.GroupVersion.String(),
+		Kind:       cluster.Kind,
+		Name:       cluster.Name,
+		UID:        cluster.UID,
+	}
+
+	if clusterapiutil.HasOwnerRef(ownerReferences, ownerReference) {
```
Comment (XudongLiuHarold) [inline addons/controllers/cpi/vspherecpiconfig_utils.go:65]: Should do the same for oracle cpi config?

File: addons/controllers/cpi/oraclecpiconfig_controller.go (hunk 1 above)
Comment (HanFa) [reply addons/controllers/cpi/oraclecpiconfig_controller.go:187]: yea its shall be fine for now, and having more consistent styke with the vspherecpiconfig controller

File: addons/controllers/cpi/vspherecpiconfig_utils.go (hunk 3 above)
Comment (HanFa) [reply addons/controllers/cpi/vspherecpiconfig_utils.go:65]: `forClusterMappedByCPIConfig` is used by both vsphere & oracle cpi config. we are finding all cpiconfigs that has owner reference to a cluster.

File: addons/controllers/cpi/oraclecpiconfig_util.go (hunk 2 above)
Comment (HanFa) [reply addons/controllers/cpi/oraclecpiconfig_util.go:35]: it is already checked https://github.com/vmware-tanzu/tanzu-framework/pull/3898/files/7bdec73b67b850c97745e9639d9bd486f0ff297a#diff-dee6d9aaf480ac1ee68633bf4b312d5e2165eb7b85cda8f67255bdaea06f3559R65 

let me put these helper functions to a new file `cpiconfig_util`

Comment (HanFa) [reply addons/controllers/cpi/oraclecpiconfig_util.go]: renamed these helper functions, logic should look more clear now

Comment (XudongLiuHarold) [reply addons/controllers/cpi/oraclecpiconfig_util.go]: 👍

**Review by XudongLiuHarold — APPROVED** [review]
LGTM

**Pull request merged**